### ISSUE TITLE
[feature] Detach weather effect particles from viewport

### ISSUE THREAD

**Issue by MarcelHB** [gemrb/gemrb#210]
**Description**
This is somewhat pedantic -- and I probably volunteer eventually to address this :wink: -- but I noticed that snow and rain effect particles are bound to the viewport, not to the area.

So if you move the view across the map, your particles follow the screen, best seen over unexplored regions. This does not look natural, even a bit cheap. Also rain particles are bit too slow.

The originals remember the particles for the visible area and some buffer around, and then redraw randomly when moving back to that area, best seen whan pausing and jumping around the map randomly.

Maybe we can do similar, or even better by parallaxing.

**Additional context**

**Comment by lynxlynxlynx** [comment]
This is handled in Particles::Draw btw. I don't think it's odd that it's sticking to the viewport, but that this hasn't been reported for any other spark.

**Comment by MarcelHB** [comment]
Interesting, I observed that in IWD2 there is this behaviour already for snow. Even more, there was a very local snowing (Targos), that was easily outlined as it was on top of some undiscovered region, about the viewports size.

As when it started to snow, the effect does exactly the opposite of what I observed in BG and rain.

**Comment by lynxlynxlynx** [comment]
Reading some old forum posts with file headers, it appears the original had (or planned to have) several intensities of rain, snow, wind (unused) and lightning.
```
CINFINITY_WEATHER_RAINLEVEL_NORAIN                      0x0000
CINFINITY_WEATHER_RAINLEVEL_LIGHTRAIN                   0x0004
CINFINITY_WEATHER_RAINLEVEL_MEDIUMRAIN                  0x0008
CINFINITY_WEATHER_RAINLEVEL_HEAVYRAIN                   0x000C
```

**Comment by MarcelHB** [comment]
I had an idea and did some sketching on Processing. Instead of painting lots of particles, we procedurally create a rain (and also snow) texture sequence that is seamless horizontally and vertically, with some parameters to adjust (# frames, # particles, speed, ...). We then have them tiled next to each other over the map, not the viewport, without the need to draw and manage individual particles.

![rain](https://user-images.githubusercontent.com/238558/127754151-5b903d05-8534-4162-8f25-7ff5cf43e3f7.gif)

**Comment by lynxlynxlynx** [comment]
Interesting idea. We'd need at least two for the rain, for the lightning highlight mentioned in #1085.

**Comment by MarcelHB** [comment]
Yeah, would need a sync with the audio and screen flashing (not sure if we do this already), but here with brightening up and reduced transparency at the slopes of its normal distribution.

![rain_lightning](https://user-images.githubusercontent.com/238558/127780496-53d0807a-b86b-4218-9a94-5d12bac9888d.gif)

**Comment by lynxlynxlynx** [comment]
We don't do any of that yet. Or shoot a real lightning now and then — in bg1 it actually could hit you and it was more likely if you carried plate mail.

**Comment by lynxlynxlynx** [comment]
Found how iwd2 decided how far to draw weather — they manually defined a bounding box depending on the resolution, quite lame:
https://github.com/search?q=repo%3Aalexbatalov%2Fiwd2-re%20stru_8E79B8&type=code
